src/org/opensolaris/opengrok/index/PendingFileCompleter.java

```diff
@@ -0,0 +1,409 @@
+import java.nio.file.StandardCopyOption;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import org.opensolaris.opengrok.logger.LoggerFactory;
+
+/**
+ * Represents a tracker of pending file deletions and renamings that can later
+ * be executed.
+ * <p>
+ * {@link PendingFileCompleter} is not generally thread-safe, as only
+ * {@link #add(org.opensolaris.opengrok.index.PendingFileRenaming)} is expected
+ * to be run in parallel; that method is thread-safe -- but only among other
+ * callers of the same method.
+ * <p>
+ * No methods are thread-safe between each other. E.g.,
+ * {@link #complete()} should only be called by a single thread after all
+ * additions of {@link PendingFileDeletion}s and {@link PendingFileRenaming}s
+ * are indicated.
+ * <p>
+ * {@link #add(org.opensolaris.opengrok.index.PendingFileDeletion)} should only
+ * be called in serial from a single thread in an isolated stage.
+ * <p>
+ * {@link #add(org.opensolaris.opengrok.index.PendingFileRenaming)}, as noted,
+ * can be called in parallel in an isolated stage.
+ */
+class PendingFileCompleter {
+
+    /**
+     * An extension that should be used as the suffix of files for
+     * {@link PendingFileRenaming} actions.
+     * <p>Value is {@code ".org_opengrok"}.
+     */
+    public static final String PENDING_EXTENSION = ".org_opengrok";
+
+    private static final Logger LOGGER =
+        LoggerFactory.getLogger(PendingFileCompleter.class);
+
+    private final Object INSTANCE_LOCK = new Object();
+
+    /**
+     * Descending path segment length comparator
+     */
+    private static final Comparator<File> DESC_PATHLEN_COMPARATOR =
+        (File f1, File f2) -> {
+            String s1 = f1.getAbsolutePath();
+            String s2 = f2.getAbsolutePath();
+            int n1 = countPathSegments(s1);
+            int n2 = countPathSegments(s2);
+            // DESC: s2 no. of segments <=> s1 no. of segments
+            int cmp = Integer.compare(n2, n1);
+            if (cmp != 0) return cmp;
+
+            // the Comparator must also be "consistent with equals", so check
+            // string contents too when (length)cmp == 0. (ASC: s1 <=> s2.)
+            cmp = s1.compareTo(s2);
+            return cmp;
+    };
+
+    private final Set<PendingFileDeletion> deletions = new HashSet<>();
+
+    private final Set<PendingFileRenaming> renames = new HashSet<>();
+
+    /**
+     * Adds the specified element to this instance's set if it is not already
+     * present.
+     * @param e element to be added to this set
+     * @return {@code true} if this instance's set did not already contain the
+     * specified element
+     */
+    public boolean add(PendingFileDeletion e) {
+        return deletions.add(e);
```

Comment:
why this is not synchronized?

Reply:
`PendingFileCompleter` is package-private but it really might be a private static class of `IndexDatabase` and only for its size did I make it a separate file. `PendingFileCompleter` accords completely with `IndexDatabase`'s algorithm where planned deletions are determined by a single thread iterating through the filesystem while simultaneously iterating through a Lucene `TermsEnum`. Synchronizing in `add(PendingFileDeletion)` would have no benefit.

Reply:
so
```java
public boolean add(PendingFileRenaming e) {
public boolean add(PendingFileDeletion e) {
```
cant happen at the same time?

Reply:
then ok

Reply:
> cant happen at the same time?

Correct — they are only called in separate stages of processing.

N.b. for the stage where `add(PendingFileRenaming)` is being executed, within the same `INSTANCE_LOCK`, the two sets tracking deletions and renamings are both modified.